lib/oelite/fetch/url.py

```diff
@@ -84,59 +86,95 @@ def localsignature(self):
         m.update(open(self.localpath, "r").read())
         return m.hexdigest()
 
-    def get_proxy(self, d):
-        proxy = d.get("http_proxy")
-        proxies = None
-        if not proxy is None:
-            proxies = {}
-            proxies["http"] = proxy
-
-            ftpproxy = d.get("ftp_proxy")
-            if ftpproxy is None:
-                proxies["ftp"] = proxy
-            else:
-                proxies["ftp"] = ftpproxy
-
-            httpsproxy = d.get("https_proxy")
-            if httpsproxy is None:
-                proxies["https"] = proxy
-            else:
-                proxies["https"] = httpsproxy
+    def get_proxies(self, d):
+        proxies = {}
+        for v in ("http_proxy", "ftp_proxy", "https_proxy"):
+            proxy = d.get(v)
+            if proxy:
+                proxies[v] = proxy
         return proxies
 
-    def get_disable_ftp_epsv(self, d):
-        val = d.get("DISABLE_FTP_EXTENDED_PASSIVE_MODE") or None
-        if val is not None:
-            if val == "0":
-                return False
-            if val == "1":
-                return True
-        return False
+    def get_passive_ftp(self, d):
+        val = d.get("DISABLE_FTP_EXTENDED_PASSIVE_MODE")
+        if val == "1":
+            return False
+        return True
 
 
 
 
-def grab(url, filename, timeout=120, retry=5, proxy=None, ftpmode=False):
+def grab(url, filename, timeout=120, retry=5, proxies=None, passive_ftp=True):
     print "Grabbing", url
 
-    if not os.path.exists(os.path.dirname(filename)):
-        os.makedirs(os.path.dirname(filename))
+    if proxies:
+        env = os.environ.copy()
+        env.update(proxies)
+    else:
+        env = None # this is the default, uses a copy of the current environment
+
+    if passive_ftp:
+        psvftp = '--passive-ftp'
+    else:
+        psvftp = '--no-passive-ftp'
+
+    d = os.path.dirname(filename)
+    f = os.path.basename(filename)
+    if not os.path.exists(d):
+        os.makedirs(d)
+
+    # Using mktemp here is perfectly fine - wget with --no-clobber and
+    # -O refuses to write to an existing file. We must download to the
+    # actual ingredient dir rather than e.g. /tmp to ensure that we
+    # can do a link(2) call without encountering EXDEV.
+    dl_tgt = tempfile.mktemp(prefix = f + ".", dir = d)
```

Comment:
Because in the 1 in 56 billion (that's 62^6, from reading tempfile.py source code) case where two racing processes end up using the same name, I don't want one wget process to silently truncate the output of another. I'd much rather the build fails, which means that it must be wget which does the O_EXCL open. We can reduce the probability further by throwing in getpid(), but PID namespaces (and network file systems) implies that doesn't make it 0.